src/main/scala/com/fulcrumgenomics/util/IntervalListSource.scala

```diff
@@ -0,0 +1,126 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.util
+
+
+import java.io.{Closeable, File, InputStream}
+
+import com.fulcrumgenomics.FgBioDef.PathToIntervals
+import com.fulcrumgenomics.commons.CommonsDef.BetterBufferedIteratorScalaWrapper
+import com.fulcrumgenomics.commons.util.StringUtil
+import htsjdk.samtools.util.{BufferedLineReader, Interval}
+import htsjdk.samtools.{SAMFileHeader, SAMSequenceDictionary, SAMTextHeaderCodec}
+
+import scala.io.Source
+
+object IntervalListSource {
+
+  /** Creates a new interval list source from a sequence of lines. */
+  def apply(lines: Seq[String]): IntervalListSource = new IntervalListSource(lines.iterator)
+
+  /** Creates a new interval list source from an iterator of lines. */
+  def apply(lines: Iterator[String]): IntervalListSource = new IntervalListSource(lines)
+
+  /** Creates a new interval list source from an input stream. */
+  def apply(stream: InputStream): IntervalListSource = new IntervalListSource(Source.fromInputStream(stream).getLines(), Some(stream))
+
+  /** Creates a new interval list source from a source. */
+  def apply(source: Source): IntervalListSource = new IntervalListSource(source.getLines(), Some(source))
+
+  /** Creates a new interval list source from a File. */
+  def apply(file: File): IntervalListSource = apply(path=file.toPath)
+
+  /** Creates a new interval list source from a Path. */
+  def apply(path: PathToIntervals): IntervalListSource = apply(Io.toInputStream(path))
+}
+
+/**
+  * Reads intervals from any text based source via a reader.  The underlying reader is closed automatically when EOF is
+  * reached.
+  */
+class IntervalListSource private(lines: Iterator[String],
+                                 private[this] val source: Option[{ def close(): Unit }] = None)
+  extends Iterator[Interval] with Closeable {
+
+  private val iter = lines.bufferBetter
+
+  private var lineNumber = 1L
+
+  // Read the header
+  val header: SAMFileHeader = {
+    val codec = new SAMTextHeaderCodec
+    val headerLines = iter.takeWhile(_.startsWith("@")).toSeq
```

Comment:
FYI I tend to prefer `toIndexedSeq` as `toSeq` will often return a lazy Stream which can be a PITA when debugging.